### PR TITLE
Add Gitlab datasource

### LINKED ISSUE #836: Add `GitLab` DataSource

### PR DESCRIPTION
**Pull request by keshav-space**
closes #836

### CODE REVIEW

File: vulntotal/datasources/gitlab.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        with tarfile.open(response.location, "r") as file_obj:
+            file_obj.extractall(extracted_location)
+        os.remove(response.location)
+        return extracted_location
+    if not speculative_execution:
+        logger.error(f"{package_slug} doesn't exist")
+    os.remove(response.location)
+
+
+def clear_download(location):
```
Comment (TG1999) [inline vulntotal/datasources/gitlab.py:107]: Add docstring

File: vulntotal/datasources/gitlab.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    ecosystem = supported_ecosystem[purl.type]
+    package_name = purl.name
+
+    if purl.type in ("maven", "composer", "golang"):
+        package_name = f"{purl.namespace}/{purl.name}"
+
+    return f"{ecosystem}/{package_name}"
+
+
+def download_subtree(package_slug: str, speculative_execution=False):
```
Comment (TG1999) [inline vulntotal/datasources/gitlab.py:91]: Add docstring

File: vulntotal/datasources/gitlab.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            "gem": "gem",
+            "golang": "go",
+            "maven": "maven",
+            "npm": "npm",
+            "nuget": "nuget",
+            "pypi": "pypi",
+        }
+
+
+def get_package_slug(purl):
```
Comment (TG1999) [inline vulntotal/datasources/gitlab.py:76]: Add docstring

File: vulntotal/datasources/gitlab.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        logger.error(f"{package_slug} doesn't exist")
+    os.remove(response.location)
+
+
+def clear_download(location):
+    if location:
+        shutil.rmtree(location)
+
+
+def get_casesensitive_slug(path, package_slug):
```
Comment (TG1999) [inline vulntotal/datasources/gitlab.py:112]: Add docstring

File: vulntotal/datasources/gitlab.py [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+                } 
+            } """,
+        }
+    ]
+    url = "https://gitlab.com/api/graphql"
+    hasnext = True
+
+    while hasnext:
+        response = requests.post(url, json=payload).json()
+        paginated_tree = response[0]["data"]["project"]["repository"]["paginatedTree"]
```
Comment (TG1999) [inline vulntotal/datasources/gitlab.py:154]: check `get_item`

File: vulntotal/datasources/gitlab.py [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+        for slug in paginated_tree["nodes"][0]["trees"]["nodes"]:
+            if slug["flatPath"].lower() == package_slug.lower():
```
Comment (TG1999) [inline vulntotal/datasources/gitlab.py:157]: check `get_item`

File: vulntotal/datasources/gitlab.py [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+            # If the namespace/subfolder contains multiple packages, then progressive transverse through folders tree
+            if package_slug.lower().startswith(slug["flatPath"].lower()):
+                return get_gitlab_style_slug(slug["flatPath"], package_slug)
+
+        payload[0]["variables"]["nextPageCursor"] = paginated_tree["pageInfo"]["endCursor"]
+        hasnext = paginated_tree["pageInfo"]["hasNextPage"]
+
+
+def parse_interesting_advisories(location, version, delete_download=False) -> Iterable[VendorData]:
+    path = Path(location)
```
Comment (TG1999) [inline vulntotal/datasources/gitlab.py:169]: Add docstring

File: vulntotal/datasources/gitlab.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+        for slug in paginated_tree["nodes"][0]["trees"]["nodes"]:
+            if slug["flatPath"].lower() == package_slug.lower():
+                return slug["flatPath"]
+
+            # If the namespace/subfolder contains multiple packages, then progressive transverse through folders tree
+            if package_slug.lower().startswith(slug["flatPath"].lower()):
+                return get_gitlab_style_slug(slug["flatPath"], package_slug)
+
+        payload[0]["variables"]["nextPageCursor"] = paginated_tree["pageInfo"]["endCursor"]
```
Comment (TG1999) [inline vulntotal/datasources/gitlab.py:164]: check `get_item`

**Review by TG1999 — CHANGES_REQUESTED** [review]

**Comment by TG1999** [comment]
Add doctsrings for all the functions and also add doctests/ unit tests for functions which have not been tested at all

File: vulntotal/datasources/gitlab.py (hunk 5 above)
Comment (keshav-space) [reply vulntotal/datasources/gitlab.py:154]: well, that won't work here.
```python
>>> from vulnerabilities.utils import get_item
>>> data = [
...   {
...     "data": {
...       "project": {
...         "repository": {
...           "paginatedTree": {
...             "pageInfo": {
...               "endCursor": "e1d88bf61d8f6e2d39bdae52121d2c19e0a86ea6",
...               "startCursor": "",
...               "hasNextPage": True
...             },
...             "nodes": [
...               {
...                 "trees": {
...                   "nodes": [
...                     {
...                       "flatPath": "pypi/AccessControl"
...                     },
...                     {
...                       "flatPath": "pypi/Acqusition"
...                     },
...                     {
...                       "flatPath": "pypi/Beaker"
...                     }
...                   ]
...                 }
...               }
...             ]
...           }
...         }
...       }
...     }
...   }
... ]
>>> get_item(data, 0, 'data', 'project', 'repository', 'paginatedTree')
dictionary must be of type `dict
```

File: vulntotal/datasources/gitlab.py (hunk 8 above)
Comment (keshav-space) [reply vulntotal/datasources/gitlab.py:164]: [https://github.com/nexB/vulnerablecode/pull/883#discussion_r963588542](url)

File: vulntotal/datasources/gitlab.py (hunk 5 above)
Comment (TG1999) [reply vulntotal/datasources/gitlab.py:154]: I meant this
```suggestion
       assert len(response) > 0
        data = response[0]
        paginated_tree = get_item(data, "data", "project", "repository", "paginatedTree")
```

File: vulntotal/datasources/gitlab.py (hunk 6 above)
Comment (TG1999) [reply vulntotal/datasources/gitlab.py:157]: Check my comment above for same.

File: vulntotal/datasources/gitlab.py (hunk 8 above)
Comment (TG1999) [reply vulntotal/datasources/gitlab.py:164]: Check my comment above for same.

**Pull request merged**